### DISCUSSION TITLE
Feature requests

### DISCUSSION THREAD

**Discussion by quiet-bear** [yds12/tarsila#12] [Ideas]
I've played around with this a bit and have enjoyed it so far (considering this is in very early stages). Here are a few of the features that I'd like to see that I didn't see on the roadmap:

- [ ] Pan camera with mouse drag
- [ ] Place color picker alongside palettes rather than toolbox
- [ ] Layer names
- [ ] Hotkeys for all the things
  - [ ] Menu actions such as saving or exporting
  - [ ] Keybindings for said hotkeys
- [ ] Support for docking the UI elements

As someone who makes a lot of pixel art, I'm willing to help out whenever I can to bring another editor into the ecosystem.

**Comment by yds12** [comment]
I'm very glad you liked the program! It's great to have some feedback from someone who actually does a lot of pixel art (I am just getting started). We are definitely in early stages, the bulk of basic functionality is available, but there is *a lot* of polish to be done, it's a rough tool for now.

> Pan camera with mouse drag

Definitely will be done, should be done by 0.2.0.

> Place color picker alongside palettes rather than toolbox

Does make sense and it is easy to implement.

> Layer names

I have considered this, or also a miniature of the layer image so that it can be easily identified. Also the image has the advantage of being automatic. But optional names are also useful indeed. I also want layers to be draggable (instead of having to click on those up/down buttons to move them). All of these are planned. (I should start fleshing out the [roadmap](https://github.com/yds12/tarsila/blob/master/ROADMAP.md) with these ideas.)

> Hotkeys for all the things

Definitely part of the roadmap. If some actions end up not having a default shortcut (if we end up having lots of menu options) then it should be configurable.

> Support for docking the UI elements

What do you mean by docking? Like when you drag it to the corner and it takes up a split of the window?

For this one we will need to check the possibilities offered by `egui`.

**Reply by quiet-bear** [reply]
> What do you mean by docking? 

Yeah, dragging the UI element onto one side of the window and having it split the screen. Either that or maybe pinning them? I like how they're floating and can be re-positioned but having some way to anchor them would be a nice addition. I don't have much experience with `egui` but I know it supports stuff like docks.

**Reply by yds12** [reply]
If egui supports it, I'm all for it. I struggled a bit with anchoring for example: the preview window is anchored to the bottom right, but in egui you either anchor the window or allow it to move (so now the preview window is not movable). Maybe we need some modification on the window decoration where we could click on a button to pin/unpin the window.

**Comment by crumblingstatue** [comment]
For docking, we could check out https://github.com/Adanos020/egui_dock
I believe emilk, the author of egui uses it in their commercial project, so it must be pretty good.
